components/CheckDocumentation/vacuum/XminHorizon.tsx

```diff
@@ -98,9 +98,48 @@ const XminHorizonGuidance: React.FunctionComponent<CheckGuidanceProps> = ({
   );
 };
 
+type HeldBackInfoType = {
+  xmin: number;
+  assigned_at: number;
+}
+
+function epochFromFullTransactionId(value: number): number {
+  return Number(BigInt(value) >> BigInt(32));
+}
+
+function xidFromFullTransactionId(value: number | null): number | null {
+  if (value == null) {
+    return null;
+  }
+  const epoch = epochFromFullTransactionId(value);
+  return fullTransactionIdMinusEpoch(value, epoch);
+}
+
+function fullTransactionIdMinusEpoch(
+  value: number,
+  epoch: number,
+): number {
+  return Number(BigInt(value) - (BigInt(epoch) << BigInt(32)));
+}
+
+const XminHeldBackInfo: React.FunctionComponent<{
+  type: string;
+  info: HeldBackInfoType;
+}> = ({ type, info }) => {
+  const epoch = epochFromFullTransactionId(info["xmin"]);
+  const xid = xidFromFullTransactionId(info["xmin"]) as number;
+
+  return (
+    <>
+      A {type} is holding back the xmin horizon at{" "}
+      <code>{`${epoch}:${xid}`}</code> (assigned at {new Date(info["assigned_at"] * 1000).toISOString()})
```

Comment:
No strong feelings. Though if you keep it, maybe this expression could be pulled up to a local in the component to keep the JSX focused on rendering.

Reply:
Reading through this again, I think the way it composes with its JSX "callers" is kind of weird. I think this would work better as just something like

```tsx
const XminHeldBackPoint: React.FunctionComponent<{
  info: HeldBackInfoType;
}> = ({ type, info }) => {
  const epoch = epochFromFullTransactionId(info["xmin"]);
  const xid = xidFromFullTransactionId(info["xmin"]) as number;
  const assignedAt = new Date(info["assigned_at"] * 1000).toISOString();
  return (
    <>
      <code>{`${epoch}:${xid}`}</code> (assigned at {assignedAt})
    </>
  )
}
```

at the cost of a bit of duplication of the "holding back the xmin horizon" text. It'd be easier to tell what the actual messages would be. Not a super-strong feeling, though.

Reply:
I was feeling a bit strange to see that "assigned at" in screenshot (thanks for adding them!), the format is not really something we use elsewhere, also as you mentioned, that info should be in the page already. I'd vote for dropping.